libraries/AC_PID/AC_P_2D.cpp

```diff
@@ -0,0 +1,44 @@
+/// @file   AC_P_2D.cpp
+/// @brief  Generic PID algorithm
+
+#include <AP_Math/AP_Math.h>
+#include "AC_P_2D.h"
+
+const AP_Param::GroupInfo AC_P_2D::var_info[] = {
+    // @Param: P
+    // @DisplayName: PI Proportional Gain
+    // @Description: P Gain which produces an output value that is proportional to the current error value
+    AP_GROUPINFO("P",    0, AC_P_2D, _kp, 0),
+    AP_GROUPEND
+};
+
+// Constructor
+AC_P_2D::AC_P_2D(float initial_p, float dt) :
+    _dt(dt)
+{
+    // load parameter values from eeprom
+    AP_Param::setup_object_defaults(this, var_info);
+
+    _kp = initial_p;
+}
+
+//  update_all - set target and measured inputs to PID controller and calculate outputs
+//  target and _error are filtered
+//  the derivative is then calculated and filtered
+//  the integral is then updated based on the setting of the limit flag
+Vector2f AC_P_2D::update_all(float &target_x, float &target_y, Vector2f measurement, float error_max, float D2_max)
```

Comment:
Changed measurement to use a const reference.

I've left the target as two separate variables because I think it would be too easy for a caller to accidentally pass in a copy of the variables instead of a reference to the variables which is required because the targets are updated.